pkg/kafka/partition/reader_service_refactor.go

```diff
@@ -0,0 +1,354 @@
+		},
+	}
+
+	var currLag time.Duration
+	for _, attempt := range attempts {
+		var err error
+
+		currLag, err = attempt()
+		if errors.Is(err, errWaitTargetLagDeadlineExceeded) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if currLag <= targetLag {
+			level.Info(logger).Log(
+				"msg", "partition reader consumed partition and current lag is lower or equal to configured target consumer lag",
+				"last_consumed_offset", s.committer.lastCommittedOffset,
+				"current_lag", currLag,
+			)
+			return nil
+		}
+	}
+
+	level.Warn(logger).Log(
+		"msg", "partition reader consumed partition and current lag is lower than configured max consumer lag but higher than target consumer lag",
+		"last_consumed_offset", s.committer.lastCommittedOffset,
+		"current_lag", currLag,
+	)
+	return nil
+}
+
+func (s *ReaderService) processNextFetchesUntilLagHonored(ctx context.Context, maxLag time.Duration, logger log.Logger, recordsChan chan<- []Record, timeSince func(time.Time) time.Duration) (time.Duration, error) {
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Second,
+		MaxRetries: 0, // Retry forever (unless context is canceled / deadline exceeded).
+	})
+	currLag := time.Duration(0)
+
+	for boff.Ongoing() {
+		// Send a direct request to the Kafka backend to fetch the partition start offset.
+		partitionStartOffset, err := s.reader.FetchPartitionOffset(ctx, kafkaStartOffset)
+		if err != nil {
+			level.Warn(logger).Log("msg", "partition reader failed to fetch partition start offset", "err", err)
+			boff.Wait()
+			continue
+		}
+
+		consumerGroupLastCommittedOffset, err := s.reader.FetchLastCommittedOffset(ctx)
+		if err != nil {
+			level.Warn(logger).Log("msg", "partition reader failed to fetch last committed offset", "err", err)
+			boff.Wait()
+			continue
+		}
+
+		// Send a direct request to the Kafka backend to fetch the last produced offset.
+		// We intentionally don't use WaitNextFetchLastProducedOffset() to not introduce further
+		// latency.
+		lastProducedOffsetRequestedAt := time.Now()
+		lastProducedOffset, err := s.reader.FetchPartitionOffset(ctx, kafkaEndOffset)
+		if err != nil {
+			level.Warn(logger).Log("msg", "partition reader failed to fetch last produced offset", "err", err)
+			boff.Wait()
+			continue
+		}
+		lastProducedOffset = lastProducedOffset - 1 // Kafka returns the next empty offset so we must subtract 1 to get the oldest written offset.
+
+		level.Debug(logger).Log("msg", "fetched latest offset information", "partition_start_offset", partitionStartOffset, "last_produced_offset", lastProducedOffset)
+
+		// Ensure there are some records to consume. For example, if the partition has been inactive for a long
+		// time and all its records have been deleted, the partition start offset may be > 0 but there are no
+		// records to actually consume.
+		if partitionStartOffset > lastProducedOffset {
```

Comment:
I wonder if this should also be `>=`. What happens if a partition is new and has had a single record written to it?

Reply:
Changing this causes `TestPartitionReader_ProcessCommits` to fail. I'd like to leave this debugging for a future PR and let this PR simply refactor the existing structure, but not the logic.